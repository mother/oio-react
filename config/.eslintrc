{
   "extends": [
      "../.eslintrc",
      "plugin:cypress/recommended"
   ],
   "env": {
     "cypress/globals": true
   },
   "plugins": ["cypress"]
}
