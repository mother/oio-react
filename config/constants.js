// =======================================================
// Breakpoint Keys
// =======================================================

const breakpointKeys = ['a', 'b', 'c', 'd', 'e', 'f']

// =======================================================
// Breakpoints with Values
// =======================================================

const breakpoints = {
   a: 475,
   b: 675,
   c: 1000,
   d: 1350,
   e: 1700,
   f: 3000
}

// =======================================================
// Breakpoints in Media Query Format
// =======================================================

const breakpointMediaQueries = {
   a: `@media (min-width: 0px) and (max-width: ${breakpoints.a}px)`,
   b: `@media (min-width: ${breakpoints.a}px) and (max-width: ${breakpoints.b}px)`,
   c: `@media (min-width: ${breakpoints.b}px) and (max-width: ${breakpoints.c}px)`,
   d: `@media (min-width: ${breakpoints.c}px) and (max-width: ${breakpoints.d}px)`,
   e: `@media (min-width: ${breakpoints.d}px) and (max-width: ${breakpoints.e}px)`,
   f: `@media (min-width: ${breakpoints.e}px)`
}

// =======================================================
// Responsive String Regex Pattern
// =======================================================

const responsiveStrFragmentRegexPattern = `(.*?)\\[([${breakpointKeys.join('')}-]+)\\]`

module.exports = {
   breakpoints,
   breakpointKeys,
   breakpointMediaQueries,
   responsiveStrFragmentRegexPattern
}
