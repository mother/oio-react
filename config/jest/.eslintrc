{
   "extends": ["../../.eslintrc"],
   "env": {
     "jest/globals": true
   },
   "plugins": ["jest"]
}
