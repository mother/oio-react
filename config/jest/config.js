module.exports = {
   rootDir: '../../',
   setupTestFrameworkScriptFile: require.resolve('./setup.js'),
   testMatch: ['<rootDir>/tests/**/*.js'],
   testPathIgnorePatterns: ['__fixtures__']
}
