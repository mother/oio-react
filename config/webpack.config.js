const path = require('path')

module.exports = {
   mode: 'production',
   entry: {
      main: './src/index.js'
   },
   output: {
      path: path.resolve(__dirname, '../dist'),
      filename: 'index.js',
      libraryTarget: 'umd'
   },
   externals: {
      react: {
         commonjs: 'react',
         commonjs2: 'react',
         amd: 'React',
         root: 'React'
      },
      'react-dom': {
         commonjs: 'react-dom',
         commonjs2: 'react-dom',
         amd: 'ReactDOM',
         root: 'ReactDOM'
      }
   },
   module: {
      rules: [
         {
            test: /\.js$/,
            exclude: /node_modules/,
            use: ['babel-loader']
         }
      ]
   },
   resolve: {
      alias: {
         react: path.resolve(__dirname, '../node_modules/react'),
         'react-dom': path.resolve(__dirname, '../node_modules/react-dom')
      }
   }
}
