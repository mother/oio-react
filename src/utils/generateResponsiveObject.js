// {
//    a: String,
//    b: String,
//    c: String,
//    d: String,
//    e: String,
//    f: String,
//    breakpointsWereSet: Boolean
// }

const breakpoints = ['a', 'b', 'c', 'd', 'e', 'f']

const generateResponsiveObject = (input) => {
   const inputStr = typeof input === 'string'
      ? input
      : String(input)

   const fragments = inputStr.match(/(.+?)\[([abcdef-]+)\]/ig)
   const result = { breakpointsWereSet: fragments !== null }

   // If no breakpoints were specified, then exit early with
   // the result for all breakpoints the same as the inputStr
   if (fragments === null) {
      for (let i = breakpoints.length - 1; i >= 0; i -= 1) {
         result[breakpoints[i]] = inputStr
      }

      return result
   }

   for (let i = fragments.length - 1; i >= 0; i -= 1) {
      const breakpointStr = fragments[i]
      const breakpointCharMatch = breakpointStr.match(/\[([abcdef-]+)\]/i)
      const breakpointChars = breakpointCharMatch[1]
      const breakpointVal = breakpointStr.replace(breakpointCharMatch[0], '').trim()

      for (let j = 0; j < breakpointChars.length; j += 1) {
         if (breakpointChars[j] === '-') {
            // TODO: HANDLE BOUNDARY ISSUES
            const startCharCode = breakpointChars.charCodeAt(j - 1)
            const endCharCode = breakpointChars.charCodeAt(j + 1)
            if (isNaN(startCharCode) || isNaN(endCharCode)) {
               throw new Error('Improperly bounded input')
            }

            for (let k = startCharCode + 1; k < endCharCode; k += 1) {
               result[String.fromCharCode(k)] = breakpointVal
            }
         } else {
            result[breakpointChars[j]] = breakpointVal
         }
      }
   }

   return result
}

export default generateResponsiveObject
