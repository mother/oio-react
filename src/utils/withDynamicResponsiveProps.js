import React, { forwardRef } from 'react'
import { breakpointKeys } from '../../config/constants'

const withDynamicResponsiveProps = iteratorFn => Component => forwardRef((props, ref) => {
   const newResponsiveProps = breakpointKeys.reduce((result, breakpointKey) => {
      const newPropsAtBreakpoint = iteratorFn(props, breakpointKey)
      Object.keys(newPropsAtBreakpoint).forEach((newPropKey) => {
         if (typeof result[newPropKey] !== 'object') {
            result[newPropKey] = { breakpointsWereSet: true }
         }

         const value = newPropsAtBreakpoint[newPropKey]
         result[newPropKey][breakpointKey] = typeof value !== 'undefined' && value !== null
            ? String(value)
            : value
      })

      return result
   }, {})

   return (
      <Component {...props} {...newResponsiveProps} ref={ref} />
   )
})

export default withDynamicResponsiveProps
