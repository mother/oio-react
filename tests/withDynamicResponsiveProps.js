import React from 'react'
import PropTypes from 'prop-types'
import { render } from 'react-testing-library'
import OIOResponsiveObjectPropType from '../src/utils/PropType'
import generateResponsiveObject from '../src/utils/generateResponsiveObject'
import withDynamicResponsiveProps from '../src/utils/withDynamicResponsiveProps'
import { breakpointKeys } from '../config/constants'

// Test Fixture

const ResponsiveText = ({ breakpoint, fontSize, fontWeight, lineHeight, value }) => (
   <span style={{
      fontSize: fontSize && fontSize[breakpoint],
      fontWeight: fontWeight && fontWeight[breakpoint],
      lineHeight: lineHeight && lineHeight[breakpoint]
   }}>
      {value}
   </span>
)

ResponsiveText.propTypes = {
   breakpoint: PropTypes.oneOf(breakpointKeys).isRequired,
   fontSize: OIOResponsiveObjectPropType.isRequired,
   fontWeight: OIOResponsiveObjectPropType,
   lineHeight: OIOResponsiveObjectPropType,
   value: PropTypes.string.isRequired
}

ResponsiveText.defaultProps = {
   fontWeight: null,
   lineHeight: null
}

// Tests

test('Will work when creating dynamic props from props at each breakpoint', () => {
   // Make the line height and font weight proportional to the font size & screen size
   const DecoratedResponsiveText = withDynamicResponsiveProps((props, breakpoint) => {
      const fontSize = parseInt(props.fontSize[breakpoint], 10)
      const { fontWeightMultiplier, lineHeightMultiplier } = props
      const baseCharCode = 'a'.charCodeAt(0)
      const currentCharCode = breakpoint.charCodeAt(0)
      // eslint-disable-next-line no-mixed-operators
      const breakpointDiff = currentCharCode - baseCharCode + 1
      return ({
         fontWeight: fontSize * fontWeightMultiplier * breakpointDiff,
         lineHeight: fontSize * lineHeightMultiplier * breakpointDiff
      })
   })(ResponsiveText)

   DecoratedResponsiveText.defaultProps = {
      fontWeightMultiplier: 1,
      fontSize: generateResponsiveObject('9[ab] 10[cd] 11[ef]'),
      lineHeightMultiplier: 0.5
   }

   const { container } = render(
      <React.Fragment>
         <DecoratedResponsiveText breakpoint="a" value="A [fw=9*1*1] [lh=9*0.5*1]" />
         <DecoratedResponsiveText breakpoint="b" value="B [fw=9*1*2] [lh=9*0.5*2]" />
         <DecoratedResponsiveText breakpoint="c" value="C [fw=10*1*3] [lh=10*0.5*3]" />
         <DecoratedResponsiveText breakpoint="d" value="D [fw=10*1*4] [lh=10*0.5*4]" />
         <DecoratedResponsiveText breakpoint="e" value="E [fw=11*1*5] [lh=11*0.5*5]" />
         <DecoratedResponsiveText breakpoint="f" value="F [fw=11*1*6] [lh=11*0.5*6]" />
      </React.Fragment>
   )

   expect(container).toMatchSnapshot()
})

test('Will not fail if a dynamic prop is only defined on certain breakpoints', () => {
   // Make the line height and font weight proportional to the font size & screen size
   const DecoratedResponsiveText = withDynamicResponsiveProps((props, breakpoint) => ({
      fontWeight: ['b', 'd', 'f'].includes(breakpoint) ? '10' : undefined,
      lineHeight: ['c', 'f'].includes(breakpoint) ? '1.1' : undefined
   }))(ResponsiveText)

   DecoratedResponsiveText.defaultProps = {
      fontSize: generateResponsiveObject('9[ab] 10[cd] 11[ef]')
   }

   const { container } = render(
      <React.Fragment>
         <DecoratedResponsiveText breakpoint="a" value="A [fw=undefined] [lh=undefined]" />
         <DecoratedResponsiveText breakpoint="b" value="B [fw=10] [lh=undefined]" />
         <DecoratedResponsiveText breakpoint="c" value="C [fw=undefined] [lh=1.1]" />
         <DecoratedResponsiveText breakpoint="d" value="D [fw=10] [lh=undefined]" />
         <DecoratedResponsiveText breakpoint="e" value="E [fw=undefined] [lh=undefined]" />
         <DecoratedResponsiveText breakpoint="f" value="F [fw=10] [lh=1.1]" />
      </React.Fragment>
   )

   expect(container).toMatchSnapshot()
})
