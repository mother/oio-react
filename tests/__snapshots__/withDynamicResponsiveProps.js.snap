// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Will not fail if a dynamic prop is only defined on certain breakpoints 1`] = `
<div>
  <span
    style="font-size: 9;"
  >
    A [fw=undefined] [lh=undefined]
  </span>
  <span
    style="font-size: 9; font-weight: 10;"
  >
    B [fw=10] [lh=undefined]
  </span>
  <span
    style="font-size: 10; line-height: 1.1;"
  >
    C [fw=undefined] [lh=1.1]
  </span>
  <span
    style="font-size: 10; font-weight: 10;"
  >
    D [fw=10] [lh=undefined]
  </span>
  <span
    style="font-size: 11;"
  >
    E [fw=undefined] [lh=undefined]
  </span>
  <span
    style="font-size: 11; font-weight: 10; line-height: 1.1;"
  >
    F [fw=10] [lh=1.1]
  </span>
</div>
`;

exports[`Will work when creating dynamic props from props at each breakpoint 1`] = `
<div>
  <span
    style="font-size: 9; font-weight: 9; line-height: 4.5;"
  >
    A [fw=9*1*1] [lh=9*0.5*1]
  </span>
  <span
    style="font-size: 9; font-weight: 18; line-height: 9;"
  >
    B [fw=9*1*2] [lh=9*0.5*2]
  </span>
  <span
    style="font-size: 10; font-weight: 30; line-height: 15;"
  >
    C [fw=10*1*3] [lh=10*0.5*3]
  </span>
  <span
    style="font-size: 10; font-weight: 40; line-height: 20;"
  >
    D [fw=10*1*4] [lh=10*0.5*4]
  </span>
  <span
    style="font-size: 11; font-weight: 55; line-height: 27.5;"
  >
    E [fw=11*1*5] [lh=11*0.5*5]
  </span>
  <span
    style="font-size: 11; font-weight: 66; line-height: 33;"
  >
    F [fw=11*1*6] [lh=11*0.5*6]
  </span>
</div>
`;
